### ISSUE TITLE
Recognise that some types always evaluate to False

### ISSUE THREAD

**Issue by DeviousStoat** [microsoft/pyright#2629]
I actually found the exact same feature request in mypy here: https://github.com/python/mypy/issues/8692 and it was added with this PR: https://github.com/python/mypy/pull/9297

```python
from typing import Literal, Union

class C:
    def __bool__(self) -> Literal[False]:
        return False

U = Union[str, C]

def f(h: U) -> str:
    if not h:
        return ""  # this could narrow down the Union to just str since C is always False

    return h  # error: Expression of type "U" cannot be assigned to return type "str"
```

So I was wondering if it could be added here as well.

Btw I am sorry if it has already been discussed here but I couldn't find anything (I'd love a link to the discussion if it has)

**Comment by erictraut** [comment]
This is an unusual pattern. The fact that it was added to mypy tells me that at least one other user found it to be useful, so it's worth considering.

That said, I have concerns about adding support for it because it's a relatively expensive check and on a hot code path in the code flow engine. I'll prototype the functionality and see if it has any measurable impact on performance. If it does, I don't think the benefits of the feature would outweigh the negative impact. If the perf impact is negligible, then I'll add it.

**Comment by erictraut** [comment]
I've implemented the feature and confirmed that its performance impact is not significant to cause any concerns. This change will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.191, which I just published. It will also be included in the next release of pylance.
